go/ai-proxy/proxy/openai/openai.go

```diff
@@ -14,8 +17,63 @@ import (
 )
 
 type OpenAIProxy struct {
-	proxy *httputil.ReverseProxy
-	token string
+	proxy        *httputil.ReverseProxy
+	tokenRotator TokenRotator
+}
+
+type TokenRotator interface {
+	GetNextToken() string
+	AddToken(token string)
+	RemoveToken(token string)
+}
+
+type RoundRobinTokenRotator struct {
+	tokens []string
+	mu     sync.RWMutex
+	index  atomic.Uint32
+}
+
+func NewRoundRobinTokenRotator(tokens []string) *RoundRobinTokenRotator {
+	tokenCpy := make([]string, len(tokens))
+	copy(tokenCpy, tokens)
+
+	return &RoundRobinTokenRotator{
+		tokens: tokenCpy,
+	}
+}
+
+func (rr *RoundRobinTokenRotator) GetNextToken() string {
+	rr.mu.RLock()
+	defer rr.mu.RUnlock()
+
+	if len(rr.tokens) == 0 {
+		return ""
+	}
+
+	currentIndex := rr.index.Add(1) - 1
+	return rr.tokens[currentIndex%uint32(len(rr.tokens))]
+}
+
+func (rr *RoundRobinTokenRotator) AddToken(token string) {
+	rr.mu.Lock()
+	defer rr.mu.Unlock()
+
+	rr.tokens = append(rr.tokens, token)
+	rr.index.Store(0) // Reset index when modifying tokens
+}
+
+func (rr *RoundRobinTokenRotator) RemoveToken(token string) {
```

Comment:
why is there a remove token function? You should just have this cycle and never mutate the underlying list

Reply:
We'd reconfigure the entire process with a new key, you can assume they're always valid if used